Passwordless login and signup

theskumar opened the pull request:
Fixes #ISSUEID

This PR is depended on #3521 

- [x] Passwordless login
- [x] Passwordless signup 
- [x] Allow user to set a password after going to profile.
- [x] Allow user to change their email even if they don't have an email set.
- [x] Allow user to add their name in the application form if name is not present in the user account.
- [x] Don't display "Dashboard" link if the user does't have permission to access to it.
- [x] Allow to use to setup 2FA without account password.
- [x] Display user content on the login screen, if configured (it is an existing feature)
- [x] If 2FA is enforced, allow the user to submit the application without setting up 2FA
- [x] Add email re-verification option to elevate, sudo mode, apart from password
- [x] Update landing page after application submission, on success it redirects now.
- [x] Update ENABLE_PUBLIC_SIGNUP and FORCE_LOGIN_FOR_APPLICATION to true by default

# Login/Signup Flow

![image](https://github.com/HyphaApp/hypha/assets/236356/91c22cb1-bd1f-4665-98e2-0350829a5807)


## Updated Login Page with Registration Enabled

![Screenshot 2023-09-13 at 07 53 06@2x](https://github.com/HyphaApp/hypha/assets/236356/424fa2f6-a519-44e3-a9ec-449815dd39b2)


## After providing the email ID

The messaging is kept neutral to hide if the user is already registered or not. The email will contain more detail, if the account exist or not.

![Screenshot 2023-09-13 at 07 57 27@2x](https://github.com/HyphaApp/hypha/assets/236356/7577447b-edf2-496e-94ca-3c98a8f2f075)

Login email copy
![Screenshot 2023-09-13 at 08 08 36@2x](https://github.com/HyphaApp/hypha/assets/236356/ab2d3e90-da2e-40ec-9f7e-bbb8d9e7235f)

## Signup

New Account Email copy
![Screenshot 2023-09-21 at 07 08 52@2x](https://github.com/HyphaApp/hypha/assets/236356/cb14350b-1cfb-4869-b863-ebbb95701089)

### Profile Page just after signup 
The user after clicking on the signup link in the email is redirect to homepage. No dashboard is available as the user doesn't have applicant role. If they click on the "profile" button they see this page with open to update profile and setup a password and enable 2FA. 

If the user decide to change the email, password is not asked if not password is set, instead an email is sent to authorize the email change.

![Screenshot 2023-09-22 at 08 50 21@2x](https://github.com/HyphaApp/hypha/assets/236356/7ed0cb97-a7ca-49fe-aef5-de8f3890db1e)


## Updated "Sudo" mode page

### For account with password

![Screenshot 2023-10-31 at 7  49 38@2x](https://github.com/HyphaApp/hypha/assets/236356/fe27ad75-7e4b-4226-89c1-ddd7a5548944)

After clicking on the "Send a confirmation code to your email" link

![Screenshot 2023-10-31 at 8  03 28@2x](https://github.com/HyphaApp/hypha/assets/236356/bf7a4098-fd0f-4d55-9850-0862da69d808)


![Screenshot 2023-10-31 at 7  51 03@2x](https://github.com/HyphaApp/hypha/assets/236356/3768e87f-d105-4d54-9cd6-2e2b8a81fa2f)


### For account without password

![Screenshot 2023-10-31 at 7  53 44@2x](https://github.com/HyphaApp/hypha/assets/236356/8c84181a-43ce-4afe-a5be-758e3f0e55f8)


## Updated disable 2FA page

It requires "Sudo" mode, instead of password now. 

![Screenshot 2023-10-31 at 7  48 01@2x](https://github.com/HyphaApp/hypha/assets/236356/2b5c8bd1-27ea-42cd-9c84-608bb351631c)

theskumar committed:
Add passwordless login landing page

theskumar committed:
Add email for login

theskumar committed:
Fix the invalid activation link template

Don't use public site template

theskumar committed:
Fix redirect url during login

theskumar committed:
Merge branch 'main' into feature/passwordless-login-signup

theskumar committed:
Fix formating

theskumar committed:
add docs

theskumar committed:
Implement signup email and token

theskumar committed:
Implement signup after clicking

Fix layout on login page

theskumar committed:
Fix tests

theskumar committed:
redirect to LOGIN_REDIRECT_URL if already authenticated

theskumar committed:
Update login form layout

theskumar committed:
login template update + no dashboard access

theskumar committed:
Do not assign Applicant role during registration

theskumar committed:
Remove register button from fund listing page

theskumar committed:
Remove Go to dashboard for those it does not apply

theskumar committed:
Don't disable the nameblock if user doesn't have name

theskumar committed:
Update "My Orgname" button of homepage" for non-applicant

theskumar committed:
Merge branch 'main' into feature/passwordless-login-signup

* main:
  Instruct prettier to ignore Sublime Text files. (#3569)
  Make it possible to control who, none/staff/staff admins, can export submissions (#3561)
  Add colour to review outcome displays and do not show "Avg. Score" if it is zero (0) (#3548)
  Add code-style checks to Github CI Actions (#3564)

theskumar committed:
Merge branch 'main' into feature/passwordless-login-signup

* main:
  Fixes after security review 2023 (#3576)
  Add slack notification for reviewers on review submit by the role reviewers (#3433)
  Add ReviewOpinion delete functionality to review sidebar (#3565)
  Fix proposal_form field disable or wrong proposal_form issue (#3570)

theskumar committed:
Update signup email copy and login/signup link timeout to 15mins

theskumar committed:
Rename ENABLE_REGISTRATION_WITHOUT_APPLICATION to ENABLE_PUBLIC_SIGNUP

theskumar committed:
Always show an option to setup 2FA

theskumar committed:
Fix doesn't ask for password if email is not changed

theskumar committed:
Remove extra 'back to account' button on disable 2FA

theskumar committed:
fixup! Remove Go to dashboard for those it does not apply

theskumar committed:
Allow user to set password if they don't have password

theskumar committed:
Update email change to use Elevate

theskumar committed:
fixup! Allow user to set password if they don't have password

theskumar committed:
Allow email to be changed even if user doesn't have password

theskumar committed:
Merge branch 'main' into feature/passwordless-login-signup

theskumar committed:
Add 2FA checks for passwordless login flow

theskumar committed:
Cosmetic updates to two factor page

theskumar committed:
If 2FA is enforced redirect to setup 2FA after signup

theskumar committed:
Add user content to login form

theskumar committed:
Comment copy update

theskumar committed:
fix test for staff with oauth can't update email

theskumar commented:
> If 2FA is enforced, allow the user to submit the application without setting up 2FA

It’s bit tricky because the way middleware for it is designed. https://github.com/HyphaApp/hypha/blob/main/hypha/apply/users/middleware.py#L40-L44
The application forms doesn’t have a url pattern, and would require a db call on each of the request that i’m trying to avoid. 

Currently trying to see if we can `mfa_exempt` a view with a decorator similar to `csrf_exempt` and catch that in the middleware before processing the view.

theskumar committed:
Enforce 2FA, direct to page after signup/login

theskumar committed:
Merge branch 'main' into feature/passwordless-login-signup

* main:
  Pin python-docx version to temporarily fix upstream issues (#3612)
  Bump postcss from 8.4.25 to 8.4.31 (#3599)

theskumar committed:
Refractor two factor middleware

- display the error message without redirecting to maintaining browser context
- performance optimization with existing early

theskumar committed:
Whitelist round and lab pages from 2FA enforcement

theskumar committed:
comment of test for project url access check

theskumar committed:
Refractor two factor middleware

- display the error message without redirecting to maintaining browser context
- performance optimization with existing early

theskumar committed:
Merge branch 'main' into feature/passwordless-login-signup

* main:
  Fix wagtail user view pagination (#3613)

theskumar committed:
Merge remote-tracking branch 'origin/main' into feature/passwordless-login-signup

* origin/main:
  Update typography of rendered rich text answers (#3601)
  Increase font-weights of page heading (#3592)
  Add status bar to projects, use same UI for submissions and projects on applicant dashboard (#3593)
  Fix the layout and styling of the upload button and file drop zone (#3595)
  Update status bar to show statuses without hover (#3586)

theskumar committed:
Implement redirection after form submission

theskumar committed:
Merge remote-tracking branch 'origin/main' into feature/passwordless-login-signup

* origin/main:
  Remove depreciation warning for mkdocs material emoji support (#3626)
  Upgrade packages with patches/security updates (#3625)
  Fix rendering of stream form fields (#3622)
  Upgrade development dependencies to latest (#3624)
  Update mkdocs version + update header color (#3623)
  Bump @babel/traverse from 7.22.8 to 7.23.2 (#3619)

theskumar committed:
Fix test for submission submit success

frjo commented:
@wes-otf To mimic OTF Live set `ENFORCE_TWO_FACTOR`, `FORCE_LOGIN_FOR_APPLICATION` and `ENABLE_REGISTRATION_WITHOUT_APPLICATION` to true

wes-otf commented:
Still going through the code with a fine tooth comb to wrap my head around it but after pulling and playing with it for a while all looks/works really good as far as I've seen! I think the flow of everything is very intuitive. Cool stuff!

@frjo just a heads up it looks like `ENABLE_REGISTRATION_WITHOUT_APPLICATION` was swapped with `ENABLE_PUBLIC_SIGNUP`. When I switched these I was able to see the "Register" option.

frjo commented on hypha/settings/base.py:
```diff
@@ -134,13 +134,9 @@
 # Forces users to log in first in order to make an application.  This is particularly useful in conjunction
-# with ENABLE_REGISTRATION_WITHOUT_APPLICATION
+# with ENABLE_PUBLIC_SIGNUP
 FORCE_LOGIN_FOR_APPLICATION = env.bool("FORCE_LOGIN_FOR_APPLICATION", False)
```
Should we rename and revert this setting?

theskumar replied:
Yes. we can do it, but can also make it backward compatible. But the enable_registration_without_application won't/doesn't make much sense.

frjo replied:
I meant the `FORCE_LOGIN_FOR_APPLICATION` setting. Switch it from default False to default True. I also think it can have a better name, especially with a new default.

frjo replied:
I like the "ENABLE_PUBLIC_SIGNUP" change.

theskumar replied:
Ah. that totally make sense! We can go for it. Can't think of a better name, but going we can potentially just remove it.

theskumar committed:
Merge remote-tracking branch 'origin/main' into feature/passwordless-login-signup

* origin/main:
  Bump werkzeug from 2.2.3 to 3.0.1 (#3630)
  Add customizable Header content field (#3604)
  Staff can upload the contract (#3614)
  Fix invoice status visibility and payment failed public message (#3616)
  Update activity feed for paf assignment action  (#3617)
  Fix draft determination disabled proposal form issue (#3618)

theskumar committed:
fixup! Implement redirection after form submission

theskumar committed:
fixup! Implement redirection after form submission

frjo commented:
For the new flow to work we need both `FORCE_LOGIN_FOR_APPLICATION` and `ENABLE_PUBLIC_SIGNUP` to the set to "True". By default they are set to "False".

1. I think there is a case for setting both of these as "True" by default. This is a big change in Hypha but it is what most everyone is using in any case.
2. I think these two settings should be right next to each other in `base.py`.

frjo commented:
Moved this out of draft/WIP. It works far to well for that.

theskumar committed:
Update 'confirm access' view to not depend on password

Also update the 2FA flows to use 'sudo' mode

Update view backup_tokens url and urlname, remove password in it

The backup token uses elevate now, and would be possible to confirm access without password

theskumar commented:
Sure. that sounds good. Just finalizing the sudo mode/confirm access without password. It will send a code/url in the email to confirm access and elevate the permission.

theskumar commented:
This is ready to testing now. Update the "Sudo" mode work without password. 

Also, updated 2FA flows to not depend on the passwords. @frjo @wes-otf @Techslammer

theskumar committed:
Merge remote-tracking branch 'origin/main' into feature/passwordless-login-signup

* origin/main:
  Add comment count to the communications tab. (#3632)

theskumar committed:
Update ENABLE_PUBLIC_SIGNUP and FORCE_LOGIN_FOR_APPLICATION to true by default

theskumar committed:
Hide login and signup button from their respective pages

theskumar committed:
Update copy and UI for disable 2FA page

theskumar commented:
Updated description with screenshots

theskumar committed:
fixup! Update ENABLE_PUBLIC_SIGNUP and FORCE_LOGIN_FOR_APPLICATION to true by default

theskumar committed:
Merge branch 'main' into feature/passwordless-login-signup

* main:
  Custom applicant dashboard header (#3587)
  Upgrade wagtail to 5x (#3627)
  Upgrade to Django 4.1.x (#3398)
  docs: update the links to images and some references to pages (#3641)
  Update the query text on all.html (#3638)
  Update to  django-3.2.23. (#3637)

frjo commented:
Will put this on test after today's testing. This will be part of a very nice Hypha 5.1 release!

theskumar committed:
Merge remote-tracking branch 'origin/main' into feature/passwordless-login-signup

* origin/main:
  Fix visibility of last item of status bar on submission detail page (#3656)
  Update test duration file and locale files. Also lock Faker to v19.x.  (#3657)

theskumar committed:
Update login with password screen

Put forgot password along-side the password field

theskumar commented:
@frjo Updated the login with password screen so that the `Forgot your password?` is just below the password field. 

Also, update the context text to be smaller/de-emphasized. 

![Screenshot 2023-11-20 at 7  17 43@2x](https://github.com/HyphaApp/hypha/assets/236356/f667802c-d6ac-4a65-9ae8-0a1606d03021)

theskumar committed:
fix check_token function

theskumar committed:
Add test for token generators

theskumar committed:
Update freezegun with time-machine

theskumar committed:
update redirect for no dashboard access
